pkg/trafficrouting/manager.go

```diff
@@ -52,142 +70,144 @@ func NewTrafficRoutingManager(c client.Client) *Manager {
+		if c.StableRevision == "" || c.CanaryRevision == "" {
+			klog.Warningf("%s stableRevision or podTemplateHash can not be empty, and wait a moment", c.Key)
+			return false, nil
+		}
+		// fetch canary service
+		err = m.Get(context.TODO(), client.ObjectKey{Namespace: c.Namespace, Name: canaryServiceName}, canaryService)
+		if err != nil && !errors.IsNotFound(err) {
+			klog.Errorf("%s get canary service(%s) failed: %s", c.Key, canaryServiceName, err.Error())
 			return false, err
+		} else if errors.IsNotFound(err) {
+			canaryService, err = m.createCanaryService(c, canaryServiceName, *stableService.Spec.DeepCopy())
+			if err != nil {
+				return false, err
+			}
 		}
-	}
 
-	// patch canary service only selector the canary pods
-	if canaryService.Spec.Selector[c.Workload.RevisionLabelKey] != canaryStatus.PodTemplateHash {
-		body := fmt.Sprintf(`{"spec":{"selector":{"%s":"%s"}}}`, c.Workload.RevisionLabelKey, canaryStatus.PodTemplateHash)
-		if err = m.Patch(context.TODO(), canaryService, client.RawPatch(types.StrategicMergePatchType, []byte(body))); err != nil {
-			klog.Errorf("rollout(%s/%s) patch canary service(%s) selector failed: %s", c.Rollout.Namespace, c.Rollout.Name, canaryService.Name, err.Error())
-			return false, err
+		// patch canary service only selector the canary pods
+		if canaryService.Spec.Selector[c.RevisionLabelKey] != c.CanaryRevision {
+			body := fmt.Sprintf(`{"spec":{"selector":{"%s":"%s"}}}`, c.RevisionLabelKey, c.CanaryRevision)
+			if err = m.Patch(context.TODO(), canaryService, client.RawPatch(types.StrategicMergePatchType, []byte(body))); err != nil {
+				klog.Errorf("%s patch canary service(%s) selector failed: %s", c.Key, canaryService.Name, err.Error())
+				return false, err
+			}
+			// update canary service time, and wait 3 seconds, just to be safe
+			c.LastUpdateTime = &metav1.Time{Time: time.Now()}
+			klog.Infof("%s patch canary service(%s) selector(%s=%s) success",
+				c.Key, canaryService.Name, c.RevisionLabelKey, c.CanaryRevision)
 		}
-		// update canary service time, and wait 3 seconds, just to be safe
-		canaryStatus.LastUpdateTime = &metav1.Time{Time: time.Now()}
-		klog.Infof("rollout(%s/%s) patch canary service(%s) selector(%s=%s) success",
-			c.Rollout.Namespace, c.Rollout.Name, canaryService.Name, c.Workload.RevisionLabelKey, canaryStatus.PodTemplateHash)
-	}
-	// patch stable service only selector the stable pods
-	if stableService.Spec.Selector[c.Workload.RevisionLabelKey] != canaryStatus.StableRevision {
-		body := fmt.Sprintf(`{"spec":{"selector":{"%s":"%s"}}}`, c.Workload.RevisionLabelKey, canaryStatus.StableRevision)
-		if err = m.Patch(context.TODO(), stableService, client.RawPatch(types.StrategicMergePatchType, []byte(body))); err != nil {
-			klog.Errorf("rollout(%s/%s) patch stable service(%s) selector failed: %s", c.Rollout.Namespace, c.Rollout.Name, stableService.Name, err.Error())
-			return false, err
+		// patch stable service only selector the stable pods
```

Comment:
patch stable service only selector the stable pods ->  patch stable service to only select stable pods